### ISSUE TITLE
Allow the use of typst for including math symbols in figures

### ISSUE THREAD

**Issue by berceanu** [MakieOrg/Makie.jl#3354]
See https://github.com/matplotlib/matplotlib/issues/26098

**Comment by jkrumbiegel** [comment]
Could you please add some more context than a link

**Comment by berceanu** [comment]
Of course. The general idea is that `matplotlib` allows users to use `LaTeX` syntax for rendering math equations in figure labels etc. (details [here](https://matplotlib.org/stable/gallery/text_labels_and_annotations/tex_demo.html)) and this functionality is heavily used for publication-quality figures. The downside is that it requires a full LaTeX distro to work, which a few GB in size normally, and is also quite slow to work with. [typst](https://github.com/typst/typst) is a new alternative, much smaller and much faster, and it is already used in Julia packages such as [polylux](https://github.com/andreasKroepelin/polylux). It is my understanding that due to the way `matplotlib` was designed, changing the typesetting engine is quite difficult, but since `makie` is much newer and still under development, it would be nice to keep this in mind as a design goal.

**Comment by sefffal** [comment]
Just a passer-by, but I think this request comes from a mistaken assumption that Makie works in a similar way to matplotlib. While Makie does support latex style syntax for specifying formula it does not require an external latex installation.
As such, the stated goal of removing the need for a Tex installation is already complete!

**Comment by avonmoll** [comment]
This feature could be facilitated by the recently released [`Typstry.jl` package](https://github.com/jakobjpeters/Typstry.jl). I've used TeX for many years and benefitted greatly by being able to use `LaTeXStrings.jl` inside of plots and figures. For several months now, I've been using Typst (almost exclusively) for all my typsetting. Having support for Typst for typesetting figure axis labels and annotations would be extremely beneficial to me so I would not have to switch to LaTeX syntax when I am generating figures.

**Comment by jakobjpeters** [comment]
> This feature could be facilitated by the recently released [Typstry.jl package](https://github.com/jakobjpeters/Typstry.jl).

Hi, I'm the author of Typstry.jl! I'd love to help out with bringing Typst to Makie, what is a high-level overview of what this requires?

**Comment by asinghvi17** [comment]
[MakieTeX.jl](https://github.com/JuliaPlots/MakieTeX.jl) just went through a pretty big refactor and now supports all kinds of vector formats (PDF, SVG, and of course TEX).  Adding Typst support (through Typstry.jl in the same way as it supports LaTeXStrings) would be pretty straightforward to me.

The final step to integration in Makie would be to add a scatter plot to the text layouting algorithm (it currently plots glyph collections and linesegments, so this isn't completely without precedent)!  Then, it would be pretty easy to integrate LatexStrings rendered via MakieTeX, and consequently some TypstString like thing as well.

**Comment by jakobjpeters** [comment]
> [MakieTeX.jl](https://github.com/JuliaPlots/MakieTeX.jl) just went through a pretty big refactor and now supports all kinds of vector formats (PDF, SVG, and of course TEX). Adding Typst support (through Typstry.jl in the same way as it supports LaTeXStrings) would be pretty straightforward to me.

Awesome! Would you want it to essentially copy the LaTeX API (`TEXDocument`, `CachedTEX`, `@Block LTeX`, etc), barring differences between the languages themselves?

> The final step to integration in Makie would be to add a scatter plot to the text layouting algorithm (it currently plots glyph collections and linesegments, so this isn't completely without precedent)!

I don't understand, but it seems to be located in src/basic_recipes/text.jl.

> Then, it would be pretty easy to integrate LatexStrings rendered via MakieTeX, and consequently some TypstString like thing as well.

Are you saying that it would support anything already in a rendered format (such as a PDF or PNG), so LaTeX and Typst could be compiled and then handled the same way (equivalent to how LaTeX is currently handled in its `_get_glyphcollection_and_linesegments` method)?

**Comment by asinghvi17** [comment]
Exactly - loading MakieTeX would ideally redefine methods in Makie to cause LaTeXStrings to render via itself instead of MathTeXEngine.  

> essentially copy the LaTeX API
Yes - the idea would be to define `TypstDocument` and `CachedTypst` (for styling reasons), along with conversion methods for e.g. `TypstStrings` (not sure what that type is called but I mean the output of `typst""` here) into full documents.  Then, there would be a rendering pipeline.  The CachedTypst struct should be identical to CachedTeX, so that we can simply widen the definition of the drawing method in the CairoMakie extension.

**Comment by SimonDanisch** [comment]
>Exactly - loading MakieTeX would ideally redefine methods in Makie to cause LaTeXStrings to render via itself instead of MathTeXEngine.

Ideally, MakieTeX won't redefine Makie methods which will change behaviour permanently just by loading it :D
I'm not sure what the ideal behaviour should be, but maybe MakieTex could export their own latex string?
Anyways, I do think it's pretty important that MakieTex can be used alongside Makie's internal MathTex based latex rendering ;)

**Comment by asinghvi17** [comment]
That's also doable - we could offer e.g. `TeX"..."` as a shorthand as well as allowing all of the document types to be rendered correctly!

**Comment by jakobjpeters** [comment]
Since MakieTeX.jl users already need to wrap their objects in things like `LTeX`, `TEXDocument`, etc, couldn't that be done internally by wrapping the `LaTeXString`s (without needing new exports)?

**Comment by asinghvi17** [comment]
This is more for the external usecase.  The idea is that we shouldn't override the native Makie rendering, but we can extend it, so a different input format would indicate to use MakieTeX rather than MathTeXEngine.  For example, this is already doable now if you wrap a `LaTeXString` in `TEXDocument`.

**Comment by lucasvalenzuela** [comment]
I would be interested in having an option to render all text through MakieTeX to allow for making plots with labels that exactly match the font and symbols used in the main text of a paper, for example. I think it would be a bit of a hurdle to manually have to wrap all regular strings and LaTeX strings just to switch over to another text rendering backend - I like the way how I can simply set `usetex=true` in the matplotlib rcParams to enable running LaTeX in the background for the math text ([see their docs](https://matplotlib.org/stable/users/explain/text/usetex.html)).

Would it be conceivable to add a "text rendering backend" type to have alongside the individual text objects? That would allow one to define it as a part of a theme and to also configure things such as a preamble for LaTeX or Typst. I was thinking of something along the lines of the following (I hope I'm getting the syntax right here for the theming part):
```julia
# tb = DefaultTextBackend() # default for the way things are currently rendered
tb = LaTeXBackend(; preamble=raw"\usepackage{newtxtext,newtxmath}")
theme = Theme(;
    Axis=(
        titlebackend=tb,
        subtitlebackend=tb,
        xlabelbackend=tb,
        xticklabelbackend=tb,
        ylabelbackend=tb,
        yticklabelbackend=tb,
        etc.
    ),
    Text=(
        textbackend=tb,
    ),
    etc.
)
```

I am not familiar with the Makie rendering structure, but I would guess it should then be possible to dispatch the text rendering according to the type of the text rendering backend, passing along all the necessary information on position, font, font size, etc.

This would mean that packages like MakieTeX could set up custom themes (similar to the already existing `theme_latexfonts()`) that set all the text backends to LaTeX or Typst, for example, and they can still be used alongside MathTeXEngine without having to change/convert the type of all the strings.

**Comment by asinghvi17** [comment]
We could just provide a method in MakieTeX that `eval`'s in the hijacking, but having to pass through an extra attribute for every text object seems like a lot of work!

**Comment by lucasvalenzuela** [comment]
Yeah, that's certainly true. Perhaps one single global text rendering object would be a better alternative to my suggestion, kind of like there is just one `usetex` in matplotlib. A user could still override it for individual plotting calls when necessary.
